### DISCUSSION TITLE
Runtime CCSID detected as -2 and is invalid error when connecting.

### DISCUSSION THREAD

**Discussion by gcraig88keys** [codefori/vscode-ibmi#2109] [IBM i]
When I connect to my IBM i in VS Code, I'm receiving this error telling me my CCSID is invalid.   My CCSID is 37, not -2.   I have reconfigured my connection to the IBM i several times.   I did delete and reinstall VS Code and the IBM i Development Pack.  The problem went away the 1st time I ran VS Code but now it's back.  Any ideas?

**Comment by sebjulliand** [comment]
Hi @gcraig88keys ,
which version of the extension are you running on?

**Reply by gcraig88keys** [reply]
The current one. V2.10.1.  I downloaded everything fresh this morning and reinstalled.

**Reply by sebjulliand** [reply]
Can you try switching to the pre-release version of Code for IBM i (2.10.7) and see if it helps? (reload the connection settings right after you installed the pre-release, just to be clean).
![image](https://github.com/codefori/vscode-ibmi/assets/11096890/33556341-f412-459f-b33d-b1fcccee9178)
![image](https://github.com/codefori/vscode-ibmi/assets/11096890/49a536ec-fb1d-461d-96a3-0e73070ed47f)

**Comment by gcraig88keys** [comment]
The message changed.  Now it's saying that system value QCCSID is invalid (-2).    That is also set to CCSID 37. 

![image](https://github.com/codefori/vscode-ibmi/assets/89139309/ba7f2083-2ff2-4320-be71-456c9a276cbe)
![image](https://github.com/codefori/vscode-ibmi/assets/89139309/3e100fbf-6c97-4fed-9fb9-235f402d0623)

**Reply by sebjulliand** [reply]
Strange! Can you use the `Connect and Reload` action again and once connected, copy/paste the `Code for IBM i output` content here?
![code for i output](https://github.com/codefori/vscode-ibmi/assets/11096890/88bc7363-2402-419f-b173-309358b914b9)
Thanks!

**Reply by gcraig88keys** [reply]
.: echo "This should be the only text!"
{
    "code": 0,
    "signal": null,
    "stdout": "This should be the only text!",
    "stderr": ""
}

.: echo $HOME && cd && test -w $HOME
{
    "code": 0,
    "signal": null,
    "stdout": "/home/OCGAR02",
    "stderr": ""
}

/home/OCGAR02: /QOpenSys/usr/bin/qsh
liblist
{
    "code": 9,
    "signal": null,
    "stdout": "",
    "stderr": ""
}

/home/OCGAR02: /QOpenSys/usr/bin/qsh
system  "CRTLIB LIB(ILEDITOR) TEXT('Code for i temporary objects. May be cleared.')"
{
    "code": 9,
    "signal": null,
    "stdout": "",
    "stderr": ""
}

/home/OCGAR02: [ -d "/tmp" ]
{
    "code": 0,
    "signal": null,
    "stdout": "",
    "stderr": ""
}

/home/OCGAR02: echo $SHELL
{
    "code": 0,
    "signal": null,
    "stdout": "/QOpenSys/usr/bin/bsh",
    "stderr": ""
}

/home/OCGAR02: /QOpenSys/usr/bin/qsh
system  "CHKOBJ OBJ(QSYS/QCPTOIMPF) OBJTYPE(*DTAARA)"
{
    "code": 9,
    "signal": null,
    "stdout": "",
    "stderr": ""
}

/home/OCGAR02: /QOpenSys/usr/bin/qsh
system  "CHKOBJ OBJ(QSYS/QCPFRMIMPF) OBJTYPE(*DTAARA)"
{
    "code": 9,
    "signal": null,
    "stdout": "",
    "stderr": ""
}

/home/OCGAR02: ls -p /usr/bin/
{
    "code": 0,
    "signal": null,
    "stdout": "ARS3WCGI\nARS400IDX\nARSADM\nARSADMIN\nARSDATE\nARSDB\nARSDOC\nARSEXPRT\nARSLOAD\nARSLOG\nARSMAINT\nARSPDOCI\nARSPDUMP\nARSSYSCR\nARSTBLSP\nRfile\najar\najar.prv\nappletviewer\napt\nar\narmsrv\narmsrvconv\nattr\nbasename\nbasename.prv\ncat\ncat.prv\ncatsplf\ncc\nchgrp\nchgrp.prv\nchmod\nchmod.prv\nchown\nchown.prv\nclrtmp\nclrtmp.prv\ncmp\ncmp.prv\ncompress\ncompress.prv\ncp\ncp.prv\ncut\ncut.prv\ndataq\ndatarea\ndate\ndate.prv\ndb2\ndb2profc\ndb2profc_trace\ndb2profp\nddns-confgen\ndelv\ndfmadmin\ndhcpd\ndhcrelay\ndig\ndirname\ndirname.prv\ndnssec-dsfromkey\ndnssec-importkey\ndnssec-keyfromlabel\ndnssec-keygen\ndnssec-revoke\ndnssec-settime\ndnssec-signzone\ndnssec-verify\ndspmsg\ndxxadm\ndxxtrc\necho\necho.prv\negrep\nenv\nenv.prv\nexpr\nexpr.prv\nextcheck\nfgrep\nfile\nfile.prv\nfind\nfind.prv\ngencat\ngetjobid\ngetjobid.prv\ngrep\ngrep.prv\nhead\nhead.prv\nhost\nhostname\nhostname.prv\nhttpapi\nhwkeytool\niconv\nid\nid.prv\nidlj\nikeyman\nipcrm\nipcs\nixlc\nixlci\njar\njarsigner\njava\njavac\njavadoc\njavah\njavap\njconsole\njdb\njdmpview\njextract\nkdestroy\nkeytab\nkeytool\nkinit\nklist\nkpasswd\nksetup\nldapadd\nldapchangepwd\nldapcompare\nldapdelete\nldapdiff\nldapexop\nldapmodify\nldapmodrdn\nldapsearch\nln\nln.prv\nlocale\nlogger\nlogger.prv\nlogname\nlogname.prv\nls\nls.prv\nmake\nmkdir\nmkdir.prv\nmkfifo\nmv\nmv.prv\nnamed\nnamed-checkconf\nnamed-checkzone\nnamed-compilezone\nnamed-journalprint\nnamed-rrchecker\nnative2ascii\nnohup\nnslookup\nnsupdate\nod\nod.prv\norbd\npack200\npax\npax.prv\npolicytool\npr\nprintenv\nprofconv\nprofdb\nprofp\nps\npwd\npwdx\npwdx.prv\nqsh\nrexec\nrexx\nrm\nrm.prv\nrmdir\nrmdir.prv\nrmic\nrmid\nrmiregistry\nrndc\nrndc-confgen\nrpm\nrsync\nschemagen\nsed\nsed.prv\nserialver\nservertool\nsetccsid\nsh\nsort\nsort.prv\nsplit\nsplit.prv\nsqlj\nstrEKM\nsys.mk\nsystem\nsysval\ntail\ntail.prv\ntar\ntee\ntee.prv\ntest\ntest.prv\ntnameserv\ntouch\ntouch.prv\ntr\ntr.prv\ntsig-keygen\nuname\nuname.prv\nuncompress\nuniq\nuniq.prv\nunpack200\nwc\nwc.prv\nwget\nwsgen\nwsimport\nxargs\nxargs.prv\nxjc\nzcat",
    "stderr": ""
}

/home/OCGAR02: ls -p /QOpenSys/pkgs/bin/
{
    "code": 2,
    "signal": null,
    "stdout": "",
    "stderr": "/QOpenSys/pkgs/bin/ not found"
}

/home/OCGAR02: ls -p /QSYS.LIB/QZDFMDB2.PGM
{
    "code": 0,
    "signal": null,
    "stdout": "/QSYS.LIB/QZDFMDB2.PGM",
    "stderr": ""
}

/home/OCGAR02: ls -p /QIBM/ProdData/IBMiDebugService/bin/startDebugService.sh
{
    "code": 0,
    "signal": null,
    "stdout": "/QIBM/ProdData/IBMiDebugService/bin/startDebugService.sh",
    "stderr": ""
}

/home/OCGAR02: ls -p /QSYS.lib/ILEDITOR.lib/GE*.PGM
{
    "code": 2,
    "signal": null,
    "stdout": "",
    "stderr": "/QSYS.lib/ILEDITOR.lib/GE*.PGM not found"
}

/home/OCGAR02: ls -p /QSYS.lib/ILEDITOR.lib/GE*.PGM
{
    "code": 2,
    "signal": null,
    "stdout": "",
    "stderr": "/QSYS.lib/ILEDITOR.lib/GE*.PGM not found"
}

/home/OCGAR02: LC_ALL=EN_US.UTF-8 system "call QSYS/QZDFMDB2 PARM('-d' '-i')"
SELECT * FROM QSYS2.ASP_INFO
{
    "code": 0,
    "signal": null,
    "stdout": "DB2>",
    "stderr": ""
}

/home/OCGAR02: LC_ALL=EN_US.UTF-8 system "call QSYS/QZDFMDB2 PARM('-d' '-i')"
select CHARACTER_CODE_SET_ID from table( QSYS2.QSYUSRINFO( USERNAME => upper('OCGAR02') ) )
{
    "code": 0,
    "signal": null,
    "stdout": "DB2>",
    "stderr": ""
}

/home/OCGAR02: /QOpenSys/usr/bin/qsh
liblist -d 
{
    "code": 9,
    "signal": null,
    "stdout": "",
    "stderr": ""
}

/home/OCGAR02: /usr/bin/attr "/QIBM/ProdData/IBMiDebugService/bin/DebugService.env" CCSID
{
    "code": 0,
    "signal": null,
    "stdout": "",
    "stderr": ""
}

/home/OCGAR02: /usr/bin/iconv -f IBM-000 -t UTF-8 "/QIBM/ProdData/IBMiDebugService/bin/DebugService.env" > /tmp/vscodetemp-O_hBI38jKZ
{
    "code": 1,
    "signal": null,
    "stdout": "",
    "stderr": "cannot open converter"
}

/home/OCGAR02: LC_ALL=EN_US.UTF-8 system "call QSYS/QZDFMDB2 PARM('-d' '-i' '-t')"
select job_name, local_port from qsys2.netstat_job_info where cast(local_port_name as VarChar(14) CCSID 37) = 'is-debug-ile' fetch first row only
/home/OCGAR02: /usr/bin/attr "/QIBM/ProdData/IBMiDebugService/bin/DebugService.env" CCSID
/home/OCGAR02: LC_ALL=EN_US.UTF-8 system "call QSYS/QZDFMDB2 PARM('-d' '-i' '-t')"
select job_name, local_port from qsys2.netstat_job_info where cast(local_port_name as VarChar(14) CCSID 37) = 'is-debug-ile' fetch first row only
/home/OCGAR02: LC_ALL=EN_US.UTF-8 system "call QSYS/QZDFMDB2 PARM('-d' '-i' '-t')"
select OS_VERSION concat '.' concat OS_RELEASE as VERSION from sysibmadm.env_sys_info
/home/OCGAR02: /usr/bin/attr "/tmp/vscodetemp-O_JQpE0BB5" CCSID
/home/OCGAR02: /usr/bin/attr "/tmp/vscodetemp-O_o6N5GGgX" CCSID
/home/OCGAR02: LC_ALL=EN_US.UTF-8 system "call QSYS/QZDFMDB2 PARM('-d' '-i' '-t')"
Call QSYS2.QCMDEXC('DSPOBJD OBJ(QSYS/) OBJTYPE(*LIB) DETAIL(*TEXTATR) OUTPUT(*OUTFILE) OUTFILE(QTEMP/LIBLIST) OUTMBR(*FIRST *ADD)');
select * from QTEMP.LIBLIST;
{
    "code": 0,
    "signal": null,
    "stdout": "",
    "stderr": ""
}

{
    "code": 0,
    "signal": null,
    "stdout": "DB2>",
    "stderr": ""
}

/home/OCGAR02: /usr/bin/iconv -f IBM-000 -t UTF-8 "/QIBM/ProdData/IBMiDebugService/bin/DebugService.env" > /tmp/vscodetemp-O_hBI38jKZ
{
    "code": 1,
    "signal": null,
    "stdout": "",
    "stderr": ""
}

{
    "code": 0,
    "signal": null,
    "stdout": "DB2>",
    "stderr": ""
}

{
    "code": 0,
    "signal": null,
    "stdout": "DB2>",
    "stderr": ""
}

{
    "code": 1,
    "signal": null,
    "stdout": "",
    "stderr": ""
}

{
    "code": 0,
    "signal": null,
    "stdout": "DB2>",
    "stderr": ""
}

/home/OCGAR02: LC_ALL=EN_US.UTF-8 system "call QSYS/QZDFMDB2 PARM('-d' '-i' '-t')"
Call QSYS2.QCMDEXC('DSPOBJD OBJ(QSYS/) OBJTYPE(*LIB) DETAIL(*TEXTATR) OUTPUT(*OUTFILE) OUTFILE(QTEMP/LIBLIST) OUTMBR(*FIRST *ADD)');
select * from QTEMP.LIBLIST;
{
    "code": 1,
    "signal": null,
    "stdout": "",
    "stderr": "cannot open converter"
}

{
    "code": 0,
    "signal": null,
    "stdout": "DB2>",
    "stderr": ""
}

/: /QOpenSys/usr/bin/qsh
system  "RUNSQLSTM SRCSTMF('/tmp/vscodetemp-O_JQpE0BB5') COMMIT(*NONE) NAMING(*SQL)"
/: /QOpenSys/usr/bin/qsh
system  "RUNSQLSTM SRCSTMF('/tmp/vscodetemp-O_o6N5GGgX') COMMIT(*NONE) NAMING(*SQL)"
{
    "code": 9,
    "signal": null,
    "stdout": "",
    "stderr": ""
}

{
    "code": 9,
    "signal": null,
    "stdout": "",
    "stderr": ""
}

**Reply by sebjulliand** [reply]
OK, I think this has to do with your SQL executions returning no results:
![image](https://github.com/codefori/vscode-ibmi/assets/11096890/63e4c55c-d01a-4031-bdc8-882925358076)

There is a specific section for this kind of issue in the documentation, with a potential fix for it (i.e. the wrong SSH server is running on your LPAR): https://codefori.github.io/docs/tips/setup/#no-results-from-sql-execution

Try the fix and then connect again.

**Comment by gcraig88keys** [comment]
Here's what I see when I run the ps -ef | grep sshd command in PASE.  
![image](https://github.com/codefori/vscode-ibmi/assets/89139309/b3ddd534-f3c5-483a-8ebe-176380dac3cd)
I don't have authority on our system to end and start the servers, and we don't want to do that in the middle of the day.   I'll talk to my administrators about doing that.

**Reply by sebjulliand** [reply]
This seems to be very similar to this issue: https://github.com/codefori/vscode-ibmi/issues/2070#issuecomment-2139551364
You're on the right path! 🤞🏻

**Reply by gcraig88keys** [reply]
I finally got around to asking the administrators to stop and start the TCP *SSHD servers on the IBM i.   That worked.   What's the magic?  How did the SSHD server get started with sshduser vs QSECOFR?  What do we need to do to make sure that it is always started correctly (with the right security)?  They probably have it programmed in the start up initial program or something. 
![image](https://github.com/codefori/vscode-ibmi/assets/89139309/be3c255f-4f3d-48cb-a907-66a5bfa374bd)

**Reply by sebjulliand** [reply]
Yep, it's definitely started by something else than `STRTCPSVR *SSHD`. It may be time to find and replace it 😄

**Reply by chrjorgensen** [reply]
@gcraig88keys I googled `ibmi sshdusr` and found this hit [Starting the SSH Daemon in a Dedicated Subsystem Environment](https://www.ibm.com/support/pages/starting-ssh-daemon-dedicated-subsystem-environment) from IBM Support. Maybe your system admin followed this IBM article?
